src/main/java/net/dries007/tfc/util/PowderKegExplosion.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Work under Copyright. Licensed under the EUPL.
+ * See the project README.md and LICENSE.txt for more information.
+ */
+
+package net.dries007.tfc.util;
+
+import java.util.*;
+
+import com.mojang.datafixers.util.Pair;
+import it.unimi.dsi.fastutil.objects.ObjectArrayList;
+import net.minecraft.core.BlockPos;
+import net.minecraft.core.particles.ParticleTypes;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.sounds.SoundEvents;
+import net.minecraft.sounds.SoundSource;
+import net.minecraft.util.Mth;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.entity.item.ItemEntity;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.level.Explosion;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.material.Material;
+import net.minecraft.world.level.storage.loot.LootContext;
+import net.minecraft.world.level.storage.loot.parameters.LootContextParams;
+import net.minecraft.world.phys.Vec3;
+
+import javax.annotation.Nullable;
+
+public class PowderKegExplosion extends Explosion
+{
+
+    private final Level level;
+    private final double x;
+    private final double y;
+    private final double z;
+    @Nullable
+    private final Entity source;
+    private final float size;
+
+    public PowderKegExplosion(Level level, Entity entity, double x, double y, double z, float size)
```

Comment:
If the field is marked `@Nullable`, so should the `Entity entity` parameter here.